File: reorderdatainlogfiles.js
let logs = ["dig1 8 1 5 1","let1 art can","dig2 3 6","let2 own kit dig","let3 art zero"]


//O(nlogn) solution that sorts the logs based on letters or digits first
//Then we sort the letter logs, and concat the finished product to 
//the digits log

let letters = []
let digits = []

for (let log of logs) {
    if (isFinite(log.split(' ')[1])) {
        digits.push(log)
    } else {
        letters.push(log)
    }
}

let sortedLetters = letters.sort((a,b) => {
    let aCut = a.slice(a.indexOf(' ') + 1)
    let bCut = b.slice(b.indexOf(' ') + 1)

    if (aCut == bCut) {
        return a > b ? 1 : -1
    } else {
        return aCut > bCut ? 1 : -1
    }
})

return sortedLetters.concat(digits)

File: reorderdatainlogfiles.java
//Java Solution

class Solution {
    public String[] reorderLogFiles(String[] logs) {
        
        Comparator<String> logComp = new Comparator<String>() {
            public int compare(String a, String b) {
                
                //Split it into two parts based on first space
                String[] aCut = a.split(" ", 2);
                String[] bCut = b.split(" ", 2);
                
                boolean isDigitA = Character.isDigit(aCut[1].charAt(0));
                boolean isDigitB = Character.isDigit(bCut[1].charAt(0));
                
                if (!isDigitA && !isDigitB) {
                    int compare = aCut[1].compareTo(bCut[1]);
                    
                    if (compare != 0) {
                        return compare;
                    }
                    
                    return aCut[0].compareTo(bCut[0]);
                }
                
                if (isDigitA && !isDigitB) {
                    return 1;
                } else if (!isDigitA && isDigitB) {
                    return -1;
                } else {
                    return 0;
                }
            }
        };
        
        Arrays.sort(logs, logComp);
        return logs;
    }
}